### ISSUE TITLE
Metrics have Nan value and cannot be published (e.g. InfluxDB, file)

### ISSUE THREAD

**Issue by IzabellaRaulin** [intelsdi-x/snap-plugin-collector-df#28]
**What happened**:
Some of metrics have NaN as a value, so snap-plugin-publisher-influxdb cannot be used to store metrics from df collector. What is more, collected metrics cannot be watched by using task watcher.

a) Publish metric to influxDB:
Received error: **Publish call error: NaN is an unsupported value for field value**
```
ID                                       NAME                                            STATE           HIT     MISS    FAIL    CREATED                 LAST FAILURE
808f536a-bd1b-4216-a1b9-16ef5fa546f9     Task-808f536a-bd1b-4216-a1b9-16ef5fa546f9       Disabled        10      0       10      4:10PM 11-28-2016       Publish call error: NaN is an unsupported value for field value

```
b) Publish metric to file by snap-plugin-publisher-file which puts data in json:
```
Publish call error: Error while marshalling metrics to JSON: json: unsupported value: NaN
```

c) Publish metric to file by mock-file publisher:

- Task is running, an exemplary published metric:
```
2016-11-28 16:17:48.020914464 +0100 CET|/intel/procfs/filesystem//boot/efi/inodes_percent_free|NaN|tags[plugin_running_on:iza-devmachine]
```

- However `task watch <task_id>` does not work - task watcher is unable to show collected metrics if one of them has NaN value.

**Steps to reproduce it**:

1. Run examplary task manifest with added section about influxDB publisher:

```
                {
                    "plugin_name": "influxdb",
                    "config": {
                        "host": "123.45.6.78",
                        "port": 8086,
                        "database": "snap",
                        "user": "myuser",
                        "password": "mypasswd"
                    }
                }
```

**Comment by IzabellaRaulin** [comment]
+ 1
snap-plugin-publisher-file cannot be used either , received the error message: `Publish call error: Error while marshalling metrics to JSON: json: unsupported value: NaN`.  (before I checked this with mock-file). 

Basing on that, I updated description and changed the title of this issue based on that

**Comment by kindermoumoute** [comment]
I think this is something we need to clarify somewhere on [plugin authoring](https://github.com/intelsdi-x/snap/blob/master/docs/PLUGIN_AUTHORING.md) or in [metric documentation](https://github.com/intelsdi-x/snap/blob/master/docs/METRICS.md). We can't use `NaN` because Go don't marshal it, and most of publisher marshal metrics. I guess the proper way to handle it is to filter and remove `NaN` metrics.

**Comment by IzabellaRaulin** [comment]
Yes, I totally agree. What is more, when some of the collected metrics have Nan value, task watcher does not work too - I update the description of this issue.

**Comment by kindermoumoute** [comment]
I took a look at the code, NaN's value come from the [syscall statfs](https://github.com/intelsdi-x/snap-plugin-collector-df/blob/master/df/plugin.go#L439). The error message show that [`stat.Ffree`](https://github.com/intelsdi-x/snap-plugin-collector-df/blob/master/df/plugin.go#L453), that means statsfs didn't error but it's still returning NaN.

You can find this behavior described [here](http://uw714doc.sco.com/en/man/html.3C/Intro.3C.html), in the "Return Values" section:
>Functions that result in a range error typically return zero for underflow or a value that will compare equal to ±HUGE_VAL for overflow. HUGE_VAL is defined in math.h. On systems that support IEEE infinity, HUGE_VAL is infinity.

> If the system supports IEEE NaN (not-a-number), functions that result in a domain error typically return NaN. Such functions may also raise one of the IEEE floating-point exceptions. On systems that do not support IEEE NaN, functions that result in a domain error typically return zero.

Now it's our role to decide the behavior we want the plugin to have. We can handle those case with [`math.IsInf`](https://golang.org/pkg/math/#IsInf) and [`math.IsNaN`](https://golang.org/pkg/math/#IsNaN).

**Comment by marcintao** [comment]
Fixed in #31 . 
Issue was caused by incautious computation of metrics - `NaN` values were results of division `0/0` for virtual filesystems.
